### PR TITLE
CI: Specify libc and libstdcxx versions in conda environments

### PR DESCRIPTION
**Pull request by peterbell10**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #121568
* #121733
* #121268
* __->__ #121556

Without this we get mismatches between the GLIBC and GLIBCXX ABI used
by conda packages vs pytorch.

### CODE REVIEW

File: .ci/docker/common/install_conda.sh [hunk 1]
```diff
@@ -58,7 +58,10 @@ fi
   # as_jenkins conda update -y -n base conda
 
   # Install correct Python version
-  as_jenkins conda create -n py_$ANACONDA_PYTHON_VERSION -y python="$ANACONDA_PYTHON_VERSION"
+  # Also ensure sysroot is using a modern GLIBC and GLIBCXX to match system compilers
```
Comment (isuruf) [inline .ci/docker/common/install_conda.sh:61]: This just updates GLIBC. Nothing to do with GLIBCXX
```suggestion
  # Also ensure sysroot is using a modern GLIBC to match system compilers
```

File: .ci/docker/common/install_conda.sh [hunk 2]
```diff
@@ -58,7 +58,14 @@ fi
   # as_jenkins conda update -y -n base conda
 
   # Install correct Python version
-  as_jenkins conda create -n py_$ANACONDA_PYTHON_VERSION -y python="$ANACONDA_PYTHON_VERSION"
+  # Also ensure sysroot is using a modern GLIBC to match system compilers
+  as_jenkins conda create -n py_$ANACONDA_PYTHON_VERSION -y\
+             python="$ANACONDA_PYTHON_VERSION" \
+             sysroot_linux-64=2.17
```
Comment (isuruf) [inline .ci/docker/common/install_conda.sh:64]: Since this script supports aarch64, this needs to change depend on `uname -m`

**Review by isuruf — APPROVED** [review]

File: .ci/docker/common/install_conda.sh [hunk 3]
```diff
@@ -57,8 +57,21 @@ fi
...
   # Install correct Python version
-  as_jenkins conda create -n py_$ANACONDA_PYTHON_VERSION -y python="$ANACONDA_PYTHON_VERSION"
+  # Also ensure sysroot is using a modern GLIBC to match system compilers
+  as_jenkins conda create -n py_$ANACONDA_PYTHON_VERSION -y\
+             python="$ANACONDA_PYTHON_VERSION" \
+             ${SYSROOT_DEP}
+
+  # libstdcxx from conda default channels are too old, we need GLIBCXX_3.4.30
+  # which is provided in libstdcxx 12 and up.
+  conda_install libstdcxx-ng=12.3.0 -c conda-forge
```
Comment (malfet) [inline .ci/docker/common/install_conda.sh:74]: Problem with this statement, is that it adds conda-forge channel, which allows the rest of the packages to be installed from conda-forge going forward. But may be this is fine

**Review by malfet — APPROVED** [review]

**Pull request closed without merging**